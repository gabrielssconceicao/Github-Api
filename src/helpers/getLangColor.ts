export const languageColors: Record<string, string> = {
  JavaScript: '#f7df1e',
  Python: '#3572A5',
  Java: '#b07219',
  'C++': '#f34b7d',
  C: '#555555',
  'C#': '#178600',
  PHP: '#4F5D95',
  TypeScript: '#2b7489',
  Swift: '#ffac45',
  Ruby: '#CC342D',
  Go: '#00ADD8',
  R: '#198CE7',
  Kotlin: '#F18E33',
  Scala: '#c22d40',
  Rust: '#dea584',
  Dart: '#00B4AB',
  Perl: '#0298c3',
  Lua: '#000080',
  Matlab: '#e16737',
  Haskell: '#5e5086',
  SQL: '#e38c00',
  HTML: '#e34c26',
  CSS: '#563d7c',
  'Objective-C': '#438eff',
  Assembly: '#6E4C13',
  Shell: '#89e051',
  PowerShell: '#012456',
  Groovy: '#e69f56',
  Vue: '#41B883',
  Dockerfile: '#384d54',
  Clojure: '#db5855',
  Elixir: '#6e4a7e',
  D: '#ba595e',
  Julia: '#a270ba',
  Arduino: '#bd79d1',
  Verilog: '#b2b7f8',
  VHDL: '#adb2cb',
  Ada: '#02f88c',
  Scheme: '#1e4aec',
  'Objective-C++': '#6866fb',
  Crystal: '#000100',
};
export const getLangColor = (lang: string = ''): string => {
  let language = '#ffffff';
  if (lang in languageColors) {
    language = languageColors[lang];
  }
  return language;
};
