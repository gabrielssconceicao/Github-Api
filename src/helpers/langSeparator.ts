import type { IRepos } from '@/types/repos';
type ILang = { name: string; count: number };
export const langSeparator = (repos: IRepos[]): ILang[] => {
  const langs: ILang[] = [];
  repos.forEach((repo) => {
    const lang = repo.language;
    const index = langs.findIndex((l) => l.name === lang);
    console.log(index);
    if (index === -1) {
      langs.push({ name: lang, count: 1 });
    } else {
      langs[index].count += 1;
    }
    console.log(langs);
  });
  return langs;
};
