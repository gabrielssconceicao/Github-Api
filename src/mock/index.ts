import { type IUser } from '@/types/user';
import { type IRepos } from '@/types/repos';

export const user: IUser = {
  name: 'Teste User',
  login: 'testeuser',
  avatar_url: 'https://avatars.githubusercontent.com/u/10590076?v=4',
  followers: 0,
  following: 5,
  location: 'Brazil',
  html_url: 'https://github.com/testeuser',
  bio: 'lorem ipsum dolor sit amet',
  public_repos: 5,
};
// JS - 3
// HTML - 2
// python - 1
// Ruby - 1
//Java - 1
//C++ - 1
//Typescript - 1
export const repos: IRepos[] = [
  {
    name: 'Teste Repo 1',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Javascript',
  },
  {
    name: 'Teste Repo 2',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'HTML',
  },
  {
    name: 'Teste Repo 3',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Python',
  },
  {
    name: 'Teste Repo 4',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Javascript',
  },
  {
    name: 'Teste Repo 5',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Javascript',
  },
  {
    name: 'Teste Repo 6',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'HTML',
  },
  {
    name: 'Teste Repo 7',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Ruby',
  },
  {
    name: 'Teste Repo 8',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Java',
  },
  {
    name: 'Teste Repo 9',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'C++',
  },
  {
    name: 'Teste Repo 10',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Typescript',
  },
];
