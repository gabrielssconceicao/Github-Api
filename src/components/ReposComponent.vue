<script setup lang="ts">
import { type IRepos } from '@/types/repos'
import { getLangColor } from '@/helpers/getLangColor'
const props = defineProps<{ repos: IRepos }>()
</script>
<template>
  <div :style="{ '--lang-color': getLangColor(props.repos.language) }" id="repo">
    <h3 class="text-center p-1 font-bold font-serif">{{ props.repos.name }}</h3>
    <p class="text-justify font-mono text-base break-words mt-3 mb-2" v-if="props.repos.description"
      aria-label="repo-description">{{
        props.repos.description }}</p>
    <div class="flex justify-between p-2">
      <p class="px-2 py-1 font-sans border-2 border-[var(--lang-color)] text-[var(--lang-color)]"
        aria-label="repo-language">{{ props.repos.language
        }}
      </p>
      <a :href="props.repos.html_url" target="_blank"
        class="border-white border-2 px-2 py-1 text-white transition-colors hover:border-[var(--lang-color)] hover:text-[var(--lang-color)]">See
        Repo</a>
    </div>
  </div>
</template>
<style scoped>
#repo {
  @apply bg-zinc-600 p-3 relative max-w-[300px] max-h-[150px] rounded-xl shadow-md before:w-2 before:absolute before:bg-[var(--lang-color)] before:top-0 before:left-0 before:bottom-0 before:rounded-tl-xl before:rounded-bl-xl md:text-xl md:w-[500px]
}
</style>
