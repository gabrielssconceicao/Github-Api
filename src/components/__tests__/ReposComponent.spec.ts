import { describe, expect, it } from 'vitest';
import { mount } from '@vue/test-utils';
import ReposComponent from '../ReposComponent.vue';
import { repos } from '@/mock';
import { getLangColor } from '@/helpers/getLangColor';

describe('ReposComponent', () => {
  it('should render correctly', () => {
    const repo = repos[0];
    const wrapper = mount(ReposComponent, {
      props: {
        repos: repo,
      },
    });
    const div = wrapper.findComponent(ReposComponent);
    expect(div.exists()).toBe(true);

    const langColor = getLangColor(repo.language);
    expect(div.attributes('style')).toBe(`--lang-color: ${langColor};`);
    expect(div.find('h3').text()).toBe(repo.name);
    expect(div.find('p[aria-label="repo-description"]').text()).toBe(
      repo.description,
    );

    expect(div.find('p[aria-label="repo-language"]').text()).toBe(
      repo.language,
    );
    expect(div.find('a').attributes('href')).toBe(repo.html_url);
    expect(div.find('a').attributes('target')).toBe('_blank');

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should not render p[aria-label="repo-description"] if description is null', () => {
    const repo = repos[0];
    const wrapper = mount(ReposComponent, {
      props: {
        repos: { ...repo, description: null },
      },
    });
    const div = wrapper.findComponent(ReposComponent);
    expect(div.exists()).toBe(true);
    expect(div.find('p[aria-label="repo-description"]').exists()).toBe(false);
    expect(wrapper.element).toMatchSnapshot();
  });
});
