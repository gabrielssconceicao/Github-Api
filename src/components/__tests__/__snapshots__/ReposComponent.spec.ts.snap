// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReposComponent > should not render p[aria-label="repo-description"] if description is null 1`] = `
<div
  data-v-7152ca0f=""
  id="repo"
  style="--lang-color: #ffffff;"
>
  <h3
    class="text-center p-1 font-bold font-serif"
    data-v-7152ca0f=""
  >
    Teste Repo 1
  </h3>
  <!--v-if-->
  <div
    class="flex justify-between p-2"
    data-v-7152ca0f=""
  >
    <p
      aria-label="repo-language"
      class="px-2 py-1 font-sans border-2 border-[var(--lang-color)] text-[var(--lang-color)]"
      data-v-7152ca0f=""
    >
      Javascript
    </p>
    <a
      class="border-white border-2 px-2 py-1 text-white transition-colors hover:border-[var(--lang-color)] hover:text-[var(--lang-color)]"
      data-v-7152ca0f=""
      href="https://github.com/testeuser/testerepo"
      target="_blank"
    >
      See Repo
    </a>
  </div>
</div>
`;

exports[`ReposComponent > should render correctly 1`] = `
<div
  data-v-7152ca0f=""
  id="repo"
  style="--lang-color: #ffffff;"
>
  <h3
    class="text-center p-1 font-bold font-serif"
    data-v-7152ca0f=""
  >
    Teste Repo 1
  </h3>
  <p
    aria-label="repo-description"
    class="text-justify font-mono text-base break-words mt-3 mb-2"
    data-v-7152ca0f=""
  >
    lorem ipsum dolor sit amet
  </p>
  <div
    class="flex justify-between p-2"
    data-v-7152ca0f=""
  >
    <p
      aria-label="repo-language"
      class="px-2 py-1 font-sans border-2 border-[var(--lang-color)] text-[var(--lang-color)]"
      data-v-7152ca0f=""
    >
      Javascript
    </p>
    <a
      class="border-white border-2 px-2 py-1 text-white transition-colors hover:border-[var(--lang-color)] hover:text-[var(--lang-color)]"
      data-v-7152ca0f=""
      href="https://github.com/testeuser/testerepo"
      target="_blank"
    >
      See Repo
    </a>
  </div>
</div>
`;
